File: exercise/c#/Day20/Day20/Domain/Yahtzee/YahtzeeCalculator.cs
namespace Day20.Domain.Yahtzee
{
    public static class YahtzeeCalculator
    {
        private const int RollLength = 5;
        private const int MinimumDie = 1;
        private const int MaximumDie = 6;

        public static int Number(int[] dice, int number)
            => Calculate(d => d.Where(die => die == number).Sum(), dice);

        public static int ThreeOfAKind(int[] dice) => CalculateNOfAKind(dice, 3);

        public static int FourOfAKind(int[] dice) => CalculateNOfAKind(dice, 4);

        public static int Yahtzee(int[] dice)
            => Calculate(d =>
                    HasNOfAKind(d, 5)
                        ? Scores.YahtzeeScore
                        : 0,
                dice);

        private static int CalculateNOfAKind(int[] dice, int n)
            => Calculate(d => HasNOfAKind(d, n) ? d.Sum() : 0, dice);

        public static int FullHouse(int[] dice)
            => Calculate(d =>
            {
                var dieFrequency = GroupDieByFrequency(d);
                return dieFrequency.ContainsValue(3) && dieFrequency.ContainsValue(2) ? Scores.HouseScore : 0;
            }, dice);

        public static int LargeStraight(int[] dice)
            => Calculate(d => d
                .OrderBy(x => x)
                .Zip(
                    d.OrderBy(x => x).Skip(1),
                    (a, b) => b - a
                ).All(diff => diff == 1)
                ? Scores.LargeStraightScore
                : 0, dice);

        public static int SmallStraight(int[] dice)
            => Calculate(d =>
            {
                var sortedDice = string.Concat(d.OrderBy(x => x).Distinct());
                return IsSmallStraight(sortedDice) ? 30 : 0;
            }, dice);

        private static bool IsSmallStraight(string diceString)
            => diceString.Contains("1234") || diceString.Contains("2345") || diceString.Contains("3456");

        private static bool HasNOfAKind(IEnumerable<int> dice, int n)
            => GroupDieByFrequency(dice).Values.Any(count => count >= n);

        public static int Chance(int[] dice) => Calculate(d => d.Sum(), dice);

        private static Dictionary<int, int> GroupDieByFrequency(IEnumerable<int> dice)
            => dice.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());

        private static int Calculate(Func<List<int>, int> compute, int[] dice)
        {
            ValidateRoll(dice);
            return compute(dice.ToList());
        }

        #region Validation

        private static void ValidateRoll(int[] dice)
        {
            if (HasInvalidLength(dice))
            {
                throw new ArgumentException("Invalid dice... A roll should contain 5 dice.");
            }

            if (ContainsInvalidDie(dice))
            {
                throw new ArgumentException("Invalid die value. Each die must be between 1 and 6.");
            }
        }

        private static bool HasInvalidLength(int[] dice) => dice is not {Length: RollLength};

        private static bool ContainsInvalidDie(IEnumerable<int> dice) => !dice.All(IsValidDie);

        private static bool IsValidDie(int die) => die is >= MinimumDie and <= MaximumDie;

        #endregion

        private static class Scores
        {
            public const int YahtzeeScore = 50;
            public const int HouseScore = 25;
            public const int LargeStraightScore = 40;
        }
    }
}

File: solution/ts/day20/src/domain/yahtzee/constrain-input/yahtzeeCalculator.ts
import {Roll} from './roll';

export class YahtzeeCalculator {
    static number = (roll: Roll, number: number): number => this.calculate(
        r => r.dice.filter(die => die === number)
            .reduce((sum, die) => sum + die, 0),
        roll
    );

    static threeOfAKind = (roll: Roll): number => this.calculateNOfAKind(roll, 3);

    static fourOfAKind = (roll: Roll): number => this.calculateNOfAKind(roll, 4);

    static yahtzee = (roll: Roll): number => this.calculate(
        r => r.dice.every(die => die === roll.dice[0])
            ? Scores.YAHTZEE_SCORE
            : 0,
        roll
    );

    static fullHouse = (roll: Roll): number => this.calculate(
        r => {
            const frequencies = Object.values(r.groupDieByFrequency());
            const hasThreeOfAKind = frequencies.includes(3);
            const hasPair = frequencies.includes(2);

            return hasThreeOfAKind && hasPair
                ? Scores.HOUSE_SCORE
                : 0;
        },
        roll
    );

    static smallStraight = (roll: Roll): number => this.calculate(
        r =>
            this.isSmallStraight(Array.from(new Set(r.dice)).sort().join(''))
                ? 30
                : 0,
        roll
    );

    static largeStraight = (roll: Roll): number => this.calculate(
        r => {
            const sortedDice = [...r.dice].sort((a, b) => a - b);
            const isSequential = sortedDice.every((die, index, arr) => index === 0 || die === arr[index - 1] + 1);

            return isSequential
                ? Scores.LARGE_STRAIGHT_SCORE
                : 0;
        },
        roll
    );

    static chance = (roll: Roll): number => this.calculate(
        r => r.dice.reduce((sum, die) => sum + die, 0),
        roll
    );

    private static calculateNOfAKind = (roll: Roll, n: number): number => this.calculate(
        r =>
            Object.values(r.groupDieByFrequency()).some(count => count >= n)
                ? r.dice.reduce((sum, die) => sum + die, 0)
                : 0,
        roll
    );
    private static isSmallStraight = (diceString: string): boolean =>
        ['1234', '2345', '3456'].some(straight => diceString.includes(straight));

    private static calculate = (compute: (roll: Roll) => number, roll: Roll): number => compute(roll);
}

const Scores = {
    YAHTZEE_SCORE: 50,
    HOUSE_SCORE: 25,
    LARGE_STRAIGHT_SCORE: 40
};